lib/Search/Index/AbstractIndexer.php

```diff
@@ -0,0 +1,352 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Affero General Public License along with
+ * this program; if not, see http://www.gnu.org/licenses or write to the Free
+ * Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301 USA.
+ *
+ * You can contact SugarCRM, Inc. headquarters at 10050 North Wolfe Road,
+ * SW2-130, Cupertino, CA 95014, USA. or at email address contact@sugarcrm.com.
+ *
+ * The interactive user interfaces in modified source and object code versions
+ * of this program must display Appropriate Legal Notices, as required under
+ * Section 5 of the GNU Affero General Public License version 3.
+ *
+ * In accordance with Section 7(b) of the GNU Affero General Public License version 3,
+ * these Appropriate Legal Notices must retain the display of the "Powered by
+ * SugarCRM" logo and "Supercharged by SuiteCRM" logo. If the display of the logos is not
+ * reasonably feasible for technical reasons, the Appropriate Legal Notices must
+ * display the words "Powered by SugarCRM" and "Supercharged by SuiteCRM".
+ */
+
+/**
+ * Created by PhpStorm.
+ * User: viocolano
+ * Date: 05/07/18
+ * Time: 11:09
+ */
+
+namespace SuiteCRM\Search\Index;
+
+use InvalidArgumentException;
+use Monolog\Handler\StreamHandler;
+use Monolog\Logger;
+use ReflectionClass;
+use SuiteCRM\Search\Index\Documentify\AbstractDocumentifier;
+use SuiteCRM\Search\Index\Documentify\JsonSerializerDocumentifier;
+
+/**
+ * This class defines common methods and fields for a search indexer.
+ *
+ * A search indexer is a component with the task of creating an index to improve search efficiency.
+ * This is usually achieved by creating a copy of the sql database in an external service.
+ *
+ * It also offers logging facilities on a separate file using Monolog, and also colored console output if configured.
+ *
+ * @see \SuiteCRM\Search\ElasticSearch\ElasticSearchIndexer
+ */
+abstract class AbstractIndexer
+{
+    /** @var bool when enabled lots output is echoed in the output stream */
+    protected $echoLogsEnabled = false;
+    /** @var bool when enabled only beans changed after the last indexing should be indexed */
+    protected $differentialIndexingEnabled = false;
+    /** @var AbstractDocumentifier determines how a bean is converted into a document */
+    protected $documentifier = null;
+    /** @var string[] The modules that have to be indexed. This should be made customisable */
+    protected $modulesToIndex = [
+        'Accounts', 'Contacts', 'Users',
+        'Opportunities', 'Leads', 'Emails',
+        'Calls', 'Meetings', 'Tasks',
+        'Surveys', 'Cases', 'Documents',
+        'Notes'
+    ];
+    /** @var Logger Monolog instance to log on a separate file */
+    protected $logger;
```

Comment:
Monolog uses handlers to do the actual logging. I refactored some code so that:
- logging happens on a separate file using Monolog StreamHandler
- logging happens on the built-in logging facility using the new SugarLoggerMonologHandler
- logging happens on CLI (stderr) using the new CliLoggerHandler

The loggier the merrier!